pkg/solana/logpoller/parser.go

```diff
@@ -0,0 +1,194 @@
+	consumedMatcher    = regexp.MustCompile(`Program \w* consumed (\d*) (.*)`)
+	logMatcher         = regexp.MustCompile(`Program log: (.*)`)
+	dataMatcher        = regexp.MustCompile(`Program data: (.*)`)
+	instructionMatcher = regexp.MustCompile(`Instruction: (.*)`)
+)
+
+type BlockData struct {
+	BlockNumber         uint64
+	BlockHash           string
+	TransactionHash     string
+	TransactionIndex    int
+	TransactionLogIndex uint
+}
+
+type ProgramLog struct {
+	BlockData
+	Text   string
+	Prefix string
+	Style  MessageStyle
+}
+
+type ProgramEvent struct {
+	BlockData
+	Prefix       string
+	FunctionName string
+	Data         string
+}
+
+type ProgramOutput struct {
+	Program      string
+	Logs         []ProgramLog
+	Events       []ProgramEvent
+	ComputeUnits uint
+	Truncated    bool
+	Failed       bool
+}
+
+func prefixBuilder(depth int) string {
+	return strings.Repeat(">", depth)
+}
+
+/*
+Program J1zQwrBNBngz26jRPNWsUSZMHJwBwpkoDitXRV95LdK4 invoke [1]
+Program log: Instruction: CreateLog
+Program data: HDQnaQjSWwkNAAAASGVsbG8sIFdvcmxkISoAAAAAAAAA // base64 encoded; borsh encoded with identifier
+Program J1zQwrBNBngz26jRPNWsUSZMHJwBwpkoDitXRV95LdK4 consumed 1477 of 200000 compute units
+Program J1zQwrBNBngz26jRPNWsUSZMHJwBwpkoDitXRV95LdK4 success
+*/
+func parseProgramLogs(logs []string) []ProgramOutput {
+	var depth int
+
+	instLogs := []ProgramOutput{}
+	lastEventIdx := -1
+
+	for _, log := range logs {
+		if strings.HasPrefix(log, "Program log:") {
+			logDataMatches := logMatcher.FindStringSubmatch(log)
+
+			if len(logDataMatches) <= 1 {
+				continue
+			}
+
+			instructionMatches := instructionMatcher.FindStringSubmatch(logDataMatches[1])
+
+			if len(instructionMatches) > 1 {
+				// is an event which should be followed by: Program data: (.*)
+				instLogs[len(instLogs)-1].Events = append(instLogs[len(instLogs)-1].Events, ProgramEvent{
+					Prefix:       prefixBuilder(depth),
+					FunctionName: instructionMatches[1],
+				})
+
+				lastEventIdx = len(instLogs[len(instLogs)-1].Events) - 1
+			} else {
+				// if contains: Instruction: (.*) this is an event and should be followed by: Program data:
+				// else this is a log
+				instLogs[len(instLogs)-1].Logs = append(instLogs[len(instLogs)-1].Logs, ProgramLog{
+					Prefix: prefixBuilder(depth),
+					Style:  MessageStyleMuted,
+					Text:   log,
+				})
+			}
+		} else if strings.HasPrefix(log, "Program data:") {
+			dataMatches := dataMatcher.FindStringSubmatch(log)
+
+			if len(dataMatches) > 1 {
+				if lastEventIdx > -1 {
+					instLogs[len(instLogs)-1].Events[lastEventIdx].Data = dataMatches[1]
+				}
+			}
+		} else if strings.HasPrefix(log, "Log truncated") {
+			instLogs[len(instLogs)-1].Truncated = true
+		} else {
+			matches := invokeMatcher.FindStringSubmatch(log)
+
+			if len(matches) > 0 {
+				if depth == 0 {
+					instLogs = append(instLogs, ProgramOutput{
+						ComputeUnits: 0,
+						Failed:       false,
+						Program:      matches[1],
```

Comment:
Can't an external program produce a log that we'll treat as a log produced by our contract?

Reply:
I'm not sure what case you're referencing. Can you expand on that?

Reply:
Let me rephrase it. Is there a guarantee that the Program's output is always base58 encoded?